precompiles/utils/src/data.rs

```diff
@@ -214,6 +226,24 @@ impl EvmData for Address {
 	}
 }
 
+//TODO @nanocryk I thought I could change this to a blanket impl like
+// impl<T: TryFrom<U256>> EvmData for T
+//
+// But the compiler gives me this error
+//
+// error[E0119]: conflicting implementations of trait `data::EvmData` for type `sp_core::H256`:
+//    --> precompiles/utils/src/data.rs:217:1
+//     |
+// 183 | impl EvmData for H256 {
+//     | --------------------- first implementation here
+// ...
+// 217 | impl<T: TryFrom<U256>> EvmData for T {
+//     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `sp_core::H256`
+//     |
+//     = note: upstream crates may add a new impl of trait `std::convert::From<sp_core::U256>` for type `sp_core::H256` in future versions
+//
+// Why should I care what they add in future versions? I should just be able to modify my code when I update the library?
+// Anyway, I needed an impl for u32 and u128. The code is very repetitive. Is this how you would have proceeded?
 impl EvmData for U256 {
 	fn read(reader: &mut EvmDataReader) -> EvmResult<Self> {
 		let range = reader.move_cursor(32)?;
```

Comment:
@librelois Maybe your Rust ninja skills can help me here too. I've had to impl this trait for several more types (u8, u32, u128) and they are all `TryFrom<U256>` so a blanket impl would be really nice.

Reply:
Each type can implement a given trait only once, in order to remove any ambiguity. This implies that you can only make a universal implementation if:

- The constraint depends on at least one trait belonging to the crate.
- The crate does not define any other implementation of this trait by foreign types.

Here, the constraint (TryFrom<U256>) does not depend on the crate (precompile-utils), so you have no control over which types meet this constraint. In the future, new types may implement TryFrom<U256>, but adding an implementation for a type is a non-breaking change, it doesn't break the current API of the type, it just adds functionality. So you can't add an implementation for the EvmData trait, because a future non-breaking change could make a given type end up with two implementations of the same trait, and the compiler won't be able to return an error because the updated dependency is supposed to have made a non-breaking change. The compiler is therefore obliged to ensure upstream that such a situation can never occur.

In your case, 2 possibilities:

1. Write a macro that repeats your impl for your different integers
2. Create a trait marker EvmDataPrecompileUtilMarker and use this trait marker as a constraint in your universal impl.

Reply:
I understand option 1 conceptually and could probably make it work after studying how to write macros. Can you say more about option 2? I'm not sure I get what you mean?

Reply:
> Can you say more about option 2? I'm not sure I get what you mean?

Some code is worth 1000 words:

```rust
pub trait EvmDataPrecompileUtilMarker {}
impl EvmDataPrecompileUtilMarker for u32 {}
impl EvmDataPrecompileUtilMarker for u128 {}

impl EvmData for T where T: TryFrom<U256> + EvmDataPrecompileUtilMarker {
    // your impl here
}
```

Reply:
Tried option 2 in 4ee815a